cowamm/kpi/competitors/curve/curve_largest_2token_pools_4232976.sql

```diff
@@ -0,0 +1,97 @@
+-- Finds all the curve pools with 2 tokens and their TVLs
+-- Parameters:
+--  {{blockchain}}: The blockchain to query
+--  {{number_of_pools}}: The number of largest pools to return
+
+with
+-- filters pools with 2 tokens
+pools as (
+    select
+        pool_address as contract_address,
+        coin0 as token0,
+        coin1 as token1,
+        mid_fee * power(10, -10) as fee
+    from curvefi_{{blockchain}}.view_pools
+    -- Curve pools can have more than 2 token, if coin2 = 0x00 it means there are 2 tokens at most
+    where coin2 = 0x0000000000000000000000000000000000000000
+),
+
+-- finds all transfers in and out of the pools to rebuild the reserves
+transfers as (
+    select
+        p.contract_address,
+        token0,
+        token1,
+        evt_block_time as block_time,
+        evt_index,
+        evt_tx_hash as tx_hash,
+        fee,
+        case
+            when t.contract_address = token0 and "from" = p.contract_address then -value
+            when t.contract_address = token0 and to = p.contract_address then value else 0
+        end as transfer0,
+        case
+            when t.contract_address = token1 and "from" = p.contract_address then -value
+            when t.contract_address = token1 and to = p.contract_address then value else 0
+        end as transfer1
+    from erc20_{{blockchain}}.evt_transfer as t
+    inner join pools as p
+        on
+            (
+                t."from" = p.contract_address
+                or t.to = p.contract_address
+            )
+            and (token0 = t.contract_address or token1 = t.contract_address)
+),
+
+-- rebuilds the reserves from the transfers
+-- ETH transfers are not considered  
+reserves as (
+    select
+        contract_address,
+        token0,
+        token1,
+        tx_hash,
+        block_time,
+        fee,
+        sum(transfer0) over (partition by contract_address order by block_time, evt_index) as reserve0,
+        sum(transfer1) over (partition by contract_address order by block_time, evt_index) as reserve1,
+        row_number() over (partition by tx_hash, contract_address order by evt_index desc) as row_num,
+        row_number() over (partition by contract_address order by block_time desc) as latest
```

Comment:
I don't understand the difference between the two. row_num is partitioned by tx_hash so it doesn't look at all at the time ordering of the transactions, does it (it just gives the latest reserve per tx)?

Isn't all we need `row_number() over (partition by contract_address order by block_time, evt_index desc)` 
 so that we can get the most up to date reserves?

Reply:
The query not only finds the largest pools, but also returns the tvl and the transfers for each tx. This avoids resimulating the reserves multiple times. I'll update the description of the query with this in mind.

latest indicates if this is the most recent reserves to get the current largest pools.
row_num gets the reserves of the pool after each block to compute the tvl at every moment in time

Reply:
I see, that's smart! Just a nit but would you mind naming this column a bit more descriptive (ie. `latest_per_tx` & `latest_per_pool`)